mono/metadata/icall.c

```diff
@@ -269,7 +269,7 @@ ves_icall_System_Array_GetValue (MonoArrayHandle arr, MonoArrayHandle indices, M
 void
 ves_icall_System_Array_SetValueImpl (MonoArrayHandle arr, MonoObjectHandle value, guint32 pos, MonoError *error)
 {
-	array_set_value_impl (arr, value, pos, FALSE, error);
+	array_set_value_impl (arr, value, pos, FALSE, TRUE, error);
```

Comment:
While we share `array_set_value_impl ` between Array.SetValue and Array.Copy/CopyConstrained the later is more relaxed around primitive type conversions